File: DemoApp/DemoApp/BaseNativeAdDetailsViewController.m
//
//  BaseNativeAdDetailsViewController.m
//  DemoApp
//
//  Created by Glispa GmbH on 19.06.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

#import "BaseNativeAdDetailsViewController.h"
#import "AppControllerManagement.h"

@interface BaseNativeAdDetailsViewController ()
@end

@implementation BaseNativeAdDetailsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)loadAds {
    [self.nativeView removeFromSuperview];
    [self prepareHeight];
}

- (void)prepareHeight {
    self.desiredHeight = 0;
}

- (void)displayAdWithAdView:(UIView *)adView {
    self.nativeView = adView;
    if (!self.nativeView.superview) {
        self.nativeView.translatesAutoresizingMaskIntoConstraints = YES;
        self.nativeContainerHeightConstraint.constant = self.desiredHeight;
        [self.adContainerView layoutIfNeeded];
        self.nativeView.frame = self.adContainerView.bounds;
        self.nativeView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self.adContainerView addSubview:self.nativeView];
        [self.adContainerView layoutIfNeeded];
        [self showBannerWithAnimation: YES];
    }
}

- (void)showBannerWithAnimation:(BOOL)animated {
    self.nativeContainerTopConstraint.constant = -self.nativeContainerHeightConstraint.constant;
    if (animated) {
        [UIView animateWithDuration:0.2 animations:^{
            [self.view layoutIfNeeded];
        }];
    } else {
        [self.view layoutIfNeeded];
    }
}

@end


@interface BaseNativeAdDetailsViewController (AppManagement) <AppControllerManagement>
@end

@implementation BaseNativeAdDetailsViewController (AppManagement)

- (void)applyUserData:(NSDictionary <NSString *, NSObject *> *)userData {
    if ([userData[@"adUnitId"] isKindOfClass:[NSString class]]) {
        self.adUnitId = (NSString *)userData[@"adUnitId"];
    }
    
    [self loadAds];
}

@end

File: DemoApp/DemoAppSwift/BaseNativeAdDetailsViewController.swift
//
//  BaseNativeAdDetailsViewController.swift
//  DemoApp
//
//  Created by Glispa GmbH on 19/05/16.
//  Copyright © 2016 Glispa GmbH. All rights reserved.
//

import UIKit
import AvocarrotNativeView

class BaseNativeAdDetailsViewController: UIViewController {

    @IBOutlet weak var adContainerView: UIView!

    public var adUnitId: String!

    @IBOutlet weak var nativeContainerHeightConstraint: NSLayoutConstraint!
    @IBOutlet weak var nativeContainerTopConstraint: NSLayoutConstraint!

    var desiredHeight: CGFloat = 0

    private var nativeView: UIView?

    override func viewDidLoad() {
        super.viewDidLoad()
    }

    func loadAds() {
        nativeView?.removeFromSuperview()
        prepareHeight()
    }

    func prepareHeight() {
        desiredHeight = 0
    }

    func displayAd(withAdView adView: UIView) {
        nativeView = adView
        if nativeView?.superview == nil {
            nativeView!.translatesAutoresizingMaskIntoConstraints = true
            nativeContainerHeightConstraint.constant = desiredHeight
            adContainerView.layoutIfNeeded()
            nativeView!.frame = adContainerView.bounds
            nativeView!.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            adContainerView.addSubview(nativeView!)
            adContainerView.layoutIfNeeded()
            showBannerWithAnimation(true)
        }
    }

    // MARK: - Private

    private func showBannerWithAnimation(_ animated: Bool) {
        nativeContainerTopConstraint.constant = -nativeContainerHeightConstraint.constant
        if animated {
            UIView.animate(withDuration: 0.2, animations: {
                self.view.layoutIfNeeded()
            })
        } else {
            view.layoutIfNeeded()
        }
    }
}

extension BaseNativeAdDetailsViewController: AppControllerManagement {
    func applyUserData(_ userData: [String:Any]) {
        if let adId = userData["adUnitId"] as? String {
            adUnitId = adId
        }

        loadAds()
    }
}
